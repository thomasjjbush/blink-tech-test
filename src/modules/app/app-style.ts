import styled from 'styled-components';

export const App = styled.main`
    display: flex;
    height: 100%;
`;
