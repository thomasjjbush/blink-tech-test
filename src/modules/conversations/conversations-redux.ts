import { getData } from '../../utils/get-data/get-data';
import { Action, Conversation, ConversationActions, Endpoints, ErrorActions, MessageActions } from '../../../types';
import { AnyAction } from 'redux';

export const loadConversations = (): Action => async (dispatch) => {
    try {
        const payload = await getData<Conversation[]>(Endpoints.CONVERSATIONS);
        dispatch({ type: ConversationActions.LOADED, payload });
    } catch (err) {
        dispatch({ type: ErrorActions.FATAL, payload: err });
    }
};

export const conversationsReducer = (state: Conversation[] = [], { payload, type }: AnyAction): Conversation[] => {
    switch (type) {
        case ConversationActions.LOADED:
            return payload;
        case MessageActions.SEND:
            return state.map((convo) =>
                convo.id === payload.conversation ? { ...convo, last_updated: payload.last_updated } : convo,
            );
        default:
            return state;
    }
};
