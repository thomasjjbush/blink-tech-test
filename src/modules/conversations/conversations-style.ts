import styled from 'styled-components';
import { StyledProps } from '../../../types';

export const Conversation = styled.button<StyledProps<{ active: boolean }>>`
    background-color: ${({ active, theme }) => (active ? theme.brightBlue : 'transparent')};
    border: 0;
    color: white;
    cursor: pointer;
    display: block;
    padding: 20px;
    text-align: left;
    width: 100%;

    &:not(:last-of-type) {
        border-bottom: solid 2px white;
    }
`;

export const Conversations = styled.div<StyledProps>`
    background-color: ${({ theme }) => theme.darkBlue};
`;

export const Name = styled.h2`
    margin: 0 0 10px;
`;
