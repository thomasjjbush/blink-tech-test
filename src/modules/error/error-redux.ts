import { AnyAction } from 'redux';
import { ErrorState, ErrorActions } from '../../../types';

export const errorReducer = (state: ErrorState = {}, { payload, type }: AnyAction): ErrorState => {
    switch (type) {
        case ErrorActions.FATAL:
            return { code: 400, fatal: true, message: payload || 'Oooops, something went wrong.' };
        default:
            return state;
    }
};
