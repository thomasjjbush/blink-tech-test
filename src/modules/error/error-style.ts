import styled from 'styled-components';

export const Error = styled.div`
    align-items: center;
    display: flex;
    flex-direction: column;
    height: 100%;
    justify-content: center;
`;
