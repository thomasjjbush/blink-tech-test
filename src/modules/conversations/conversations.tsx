import React, { FC, ReactElement, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { loadConversations } from './conversations-redux';
import { DateAndTime } from '../../components/date-and-time/date-and-time';
import { Conversation, ConversationsProps as Props, Store } from '../../../types';
import * as Styled from './conversations-style';

export const Conversations: FC<Props> = ({ conversation, setConversation }: Props): ReactElement => {
    const dispatch = useDispatch();
    const conversations = useSelector<Store, Conversation[]>((state) => state.conversations);

    useEffect(() => {
        (async () => {
            await dispatch(loadConversations());
        })();
    }, []);

    return (
        <Styled.Conversations>
            {conversations
                .sort((a, b) => new Date(b.last_updated).valueOf() - new Date(a.last_updated).valueOf())
                .map(({ id, name, last_updated }) => (
                    <Styled.Conversation
                        active={conversation === id}
                        key={id}
                        onClick={() => setConversation(id)}
                        type="button"
                    >
                        <Styled.Name>{name}</Styled.Name>
                        <DateAndTime>{last_updated}</DateAndTime>
                    </Styled.Conversation>
                ))}
        </Styled.Conversations>
    );
};
