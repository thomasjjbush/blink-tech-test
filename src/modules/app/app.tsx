import React, { FC, lazy, ReactElement, Suspense, useState } from 'react';
import { useSelector } from 'react-redux';
import { Loading } from '../../components/loading/loading';
import { Conversations } from '../conversations/conversations';
import { Error } from '../error/error';
import { ErrorState, Store } from '../../../types';
import * as Styled from './app-style';

const Messages = lazy(() => import('../messages/messages').then((module) => ({ default: module.Messages })));

export const App: FC = (): ReactElement => {
    const error = useSelector<Store, ErrorState>((state) => state.error);
    const [conversation, setConversation] = useState<string>(null);

    if (error.fatal) return <Error code={error.code} message={error.message} />;

    return (
        <Styled.App>
            <Conversations conversation={conversation} setConversation={setConversation} />
            {conversation && (
                <Suspense fallback={<Loading />}>
                    <Messages conversation={conversation} />
                </Suspense>
            )}
        </Styled.App>
    );
};
