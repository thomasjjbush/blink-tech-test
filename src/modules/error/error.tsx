import React, { FC, ReactElement } from 'react';
import * as Styled from './error-style';
import { ErrorState as Props } from '../../../types';

export const Error: FC<Props> = ({ code, message }: Props): ReactElement => (
    <Styled.Error>
        <h1>{code}</h1>
        <p>{message}</p>
    </Styled.Error>
);
