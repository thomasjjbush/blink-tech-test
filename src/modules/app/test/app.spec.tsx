import React from 'react';
import { shallow } from 'enzyme';
import { useSelector } from 'react-redux';
import { App } from '../app';
import { Conversations } from '../../conversations/conversations';
import { Error } from '../../error/error';

jest.mock('react-redux', () => ({ useSelector: jest.fn((cb) => cb({ error: {} })) }));

describe('App', () => {
    it('should match snapshot when conversation has not been selected', () => {
        expect(shallow(<App />)).toMatchSnapshot();
    });

    it('should match snapshot when conversation has been selected', () => {
        const wrapper = shallow(<App />);
        wrapper.find(Conversations).prop('setConversation')('conversation-id');

        expect(wrapper).toMatchSnapshot();
    });

    it('should return fatal error UI if fatal error exists in store', () => {
        (useSelector as jest.Mock).mockImplementationOnce((cb) => cb({ error: { code: 6, fatal: true, message: '' } }));
        const wrapper = shallow(<App />).find(Error);

        expect(wrapper.prop('code')).toBe(6);
        expect(wrapper.prop('message')).toBe('');
    });
});
