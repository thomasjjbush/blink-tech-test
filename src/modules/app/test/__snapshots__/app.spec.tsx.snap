// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot when conversation has been selected 1`] = `
<styled.main
  styles="
    display: flex;
    height: 100%;
"
>
  <Conversations
    conversation="conversation-id"
    setConversation={[Function]}
  />
  <Suspense
    fallback={<Loading />}
  >
    <lazy
      conversation="conversation-id"
    />
  </Suspense>
</styled.main>
`;

exports[`App should match snapshot when conversation has not been selected 1`] = `
<styled.main
  styles="
    display: flex;
    height: 100%;
"
>
  <Conversations
    conversation={null}
    setConversation={[Function]}
  />
</styled.main>
`;
